src/Env.ml

```diff
@@ -31,19 +34,29 @@ struct
   (** Extend environment with a key and the corresponding value *)
   let extend env k v = (k, v) :: env
 
-  (** Find the value corresponding to a key identifier *)
+  (** Find the value corresponding to a key identifier
+      NOTE: effectively type of error (EnvUnboundRegularVarError or
+            EnvUnboundModalVariableError) depends on type of ID, but ocaml can't check it.
+   *)
   let lookup env k =
     let typ_o = List.Assoc.find env k ~equal:Key.equal in
     match typ_o with
     | Some t -> return t
     | None ->
-        let var_name = [%sexp_of: Key.t] k |> Sexp.to_string in
-        Result.fail
-        @@ `EnvUnboundVariableError
-             ( var_name,
-               [%string
-                 "$(var_name) is not found in the  $(Key.context_kind) \
-                  environment!"] )
+        let var_name = Key.to_string k in
+        let message =
+          [%string
+            "$(var_name) is not found in the $(Key.context_kind) environment!"]
+        in
+        let error =
+          if [%equal: string] Key.context_kind Id.R.context_kind then
+            `EnvUnboundRegularVarError (var_name, message)
+          else if [%equal: string] Key.context_kind Id.M.context_kind then
+            `EnvUnboundModalVarError (var_name, message)
+          else assert false
+          (* Impossible *)
```

Comment:
Could we please revisit this design decision? This place does not need to know about all the possible instantiations.

Reply:
Fixed, but still not so well

Reply:
LGTM. Maybe will think of a better solution later.